### DISCUSSION TITLE
Contour plot of a function with complex values

### DISCUSSION THREAD

**Discussion by edelopo** [pyvista/pyvista#3706] [Q&A] [answered]
I have a function f: R^3 -> C, that is, a function on space with complex values. I want to plot the isosurface of points that get mapped to the positive real axis. I had the idea of using Im(f) = 0, and then doing .boolean_difference() with Re(f) = 0 to get either of the two sides of the axis (I don't care which one I get). However, this is not working and the difference is doing kind of a union. Any idea on how could I accomplish something like this?

EDIT: I guess that the behaviour I'm looking for is analogous to the RegionFunction of Mathematica, see https://mathematica.stackexchange.com/a/87502. Is there anything like that in pyvista?

**Comment by MatthewFlamm** [comment, marked as answer]
Is there a minimal example you can share?  I'm not quite following.  One way of reading your statement is that you want to only keep the part of a mesh based on data (in semi-pseudo code):

```py
mesh # UnstructuredGrid or similar 3D space filling mesh
mesh['real_data'] = real_part_of_data
mesh['imaginary_data'] = imaginary_part_of_data
mesh.contour('real_data')  # isosurface
mesh.clip('real_data')  # clip away either positive or negative values
```

**Reply by edelopo** [reply]
For sure, here you have an example of what I'm trying:
```
import numpy as np
import pyvista as pv

contours = []

#%% Data
x, y, z = np.pi*np.mgrid[-2:2:51j, -2:2:51j, -2:2:51j]
real_part = x
imaginary_part = y

im = imaginary_part.flatten()
re = real_part.flatten()

grid = pv.StructuredGrid(x, y, z)
grid['im'] = im
grid['re'] = re
uncut_mesh = grid.contour([0], scalars='im')
cutting_mesh = grid.contour([0], scalars='re')
final_mesh = uncut_mesh.boolean_difference(cutting_mesh)

#%% Visualization
pv.set_plot_theme('default')
p = pv.Plotter()
p.add_mesh(final_mesh, style='surface', color='blue')
p.show(interactive_update=False)
```
In this example, the set im=0 is the plane y=0 and the set re=0 is the plane x=0. What I would expect to get when I do the boolean difference is the half plane y>0 (or y<0), but the result is the union of two half planes, as you can see in the picture.

![image](https://user-images.githubusercontent.com/18720965/207824045-3e2c6f3f-7a15-482b-b57d-3cb0e9ba13f5.png)

**Reply by edelopo** [reply]
I am trying you approach, but `grid.clip('re')` throws a `KeyError`. How do I use clip?

**Reply by edelopo** [reply]
Nevermind, `clip_surface` is precisely what I was looking for. It works now!
